repoze/lru/__init__.py

```diff
@@ -1,23 +1,79 @@
 """ LRU caching class and decorator """
 from __future__ import with_statement
+from abc import ABCMeta, abstractmethod
 
 import threading
 import time
 import uuid
+from warnings import warn
 
 
 _MARKER = object()
 # By default, expire items after 2**60 seconds. This fits into 64 bit
 # integers and is close enough to "never" for practical purposes.
 _DEFAULT_TIMEOUT = 2 ** 60
 
+_WARN = True
 
-class LRUCache(object):
+
+def disable_warnings():
+    global _WARN
+    _WARN = False
+
+
+def enable_warnings():
+    global _WARN
+    _WARN = True
+
+
+class Cache(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def clear(self):
+        pass
+
+    @abstractmethod
+    def get(self, key, default=None):
+        pass
+
+    @abstractmethod
+    def put(self, key, val):
+        pass
+
+    @abstractmethod
+    def invalidate(self, key):
+        pass
+
+
+class UnboundedCache(Cache):
+    """
+    a simple unbounded cache backed by a dictionary
+    """
+
+    def __init__(self):
+        self._data = dict()
+
+    def get(self, key, default=None):
+        return self._data.get(key, default)
+
+    def clear(self):
+        self._data.clear()
+
+    def invalidate(self, key):
+        pass
```

Comment:
I don't see how this can be correct:  invalidating an item *must* cause the previously-cached value to be cleared.